### PR TITLE
[DROOLS-6912] prop with method invocation is not recognized in modify…

### PR DESCRIPTION
**Pull request by tkobayas**
… block with executable model

**Ports** 
This PR is for 7.x.
for main -> https://github.com/kiegroup/drools/pull/4338

**JIRA**: 
https://issues.redhat.com/browse/DROOLS-6912

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>

* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`

* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by tkobayas** [comment]
GHA windows-latest - Java 11: ... Need rebase
```
2022-04-21T09:37:37.9839968Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project drools-mvel-compiler: Compilation failure
2022-04-21T09:37:37.9840982Z [ERROR] /D:/a/drools/drools/kiegroup_drools/drools-model/drools-mvel-compiler/src/test/java/org/drools/mvelcompiler/MvelCompilerTest.java:[765,60] cannot find symbol
2022-04-21T09:37:37.9841575Z [ERROR]   symbol:   method containsInAnyOrder(java.lang.String)
2022-04-21T09:37:37.9842044Z [ERROR]   location: class org.drools.mvelcompiler.MvelCompilerTest
2022-04-21T09:37:37.9842490Z [ERROR] -> [Help 1]
2022-04-21T09:37:37.9843437Z org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project drools-mvel-compiler: Compilation failure
2022-04-21T09:37:37.9858256Z /D:/a/drools/drools/kiegroup_drools/drools-model/drools-mvel-compiler/src/test/java/org/drools/mvelcompiler/MvelCompilerTest.java:[765,60] cannot find symbol
2022-04-21T09:37:37.9858930Z   symbol:   method containsInAnyOrder(java.lang.String)
2022-04-21T09:37:37.9859416Z   location: class org.drools.mvelcompiler.MvelCompilerTest
```

**Comment by tkobayas** [comment]
GHA : windows-latest - Java 11: non-exec-model test so not related to this PR. Probably flaky. 
```
2022-04-21T16:44:01.7601926Z [ERROR] Tests run: 238, Failures: 1, Errors: 0, Skipped: 2, Time elapsed: 63.577 s <<< FAILURE! - in org.drools.compiler.integrationtests.CepEspTest
2022-04-21T16:44:01.7602672Z [ERROR] testDeserializationWithExpiringEventAndAccumulate[KieBase type=STREAM_IDENTITY](org.drools.compiler.integrationtests.CepEspTest)  Time elapsed: 1.327 s  <<< FAILURE!
2022-04-21T16:44:01.7603231Z java.lang.AssertionError: expected:<[2.0, 5.0]> but was:<[3.0, 6.0]>
```

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java [hunk 1]
```diff
@@ -220,25 +220,48 @@ private Statement addScopeToMethodCallExpr(PreprocessPhaseResult result, Express
...
                 }
-
-                Expression enclosed = new EnclosedExpr(scope);
-                rootMcExpr.setScope(enclosed);
-
-                if (scope.isNameExpr() || scope instanceof DrlNameExpr) { // some classes such "AtomicInteger" have a setter called "set"
-                    result.addUsedBinding(printNode(scope));
-                }
-
-                return new ExpressionStmt(mcExpr);
```
Comment (tkobayas) [inline drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java]: Originally, `PreprocesPhase` only sets scope for MethodCallExpr. (e.g. `modify($p) { toString() };` ->  `{ $p.toString(); }`). This PR can pick up the root property and set scope. (e.g. `modify($p) { addresses.clear() };` -> `{ $p.getAddresses().clear(); }`)

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java [hunk 2]
```diff
@@ -220,25 +220,48 @@ private Statement addScopeToMethodCallExpr(PreprocessPhaseResult result, Express
...
+                    Expression enclosed = new EnclosedExpr(scope);
+                    rootMcExpr.setScope(enclosed);
+
+                    if (scope.isNameExpr() || scope instanceof DrlNameExpr) { // some classes such "AtomicInteger" have a setter called "set"
+                        result.addUsedBinding(printNode(scope));
+                    }
+
+                    return new ExpressionStmt(mcExpr);
+                } else if (rootExpr instanceof DrlNameExpr) {
+                    if (isSameDrlName(rootExpr, scope)) {
```
Comment (mariofusco) [inline drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java]: Nitpick: I don't like an empty `if` branch, with something in the `else` one. Please get rid of it and just do `if (!isSameDrlName(rootExpr, scope))` ...

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java [hunk 3]
```diff
@@ -220,25 +220,48 @@ private Statement addScopeToMethodCallExpr(PreprocessPhaseResult result, Express
...
     }
 
-    private MethodCallExpr findRootScope(MethodCallExpr mcExpr) {
+    private boolean isSameDrlName(Expression rootExpr, Expression scope) {
+        if (rootExpr instanceof DrlNameExpr && scope instanceof DrlNameExpr) {
+            if (((DrlNameExpr) rootExpr).getName().equals(((DrlNameExpr) scope).getName())) {
+                return true;
+            }
+        }
+        return false;
```
Comment (mariofusco) [inline drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java]: This is a bit convoluted, I believe that the following will do the same:
```
return rootExpr instanceof DrlNameExpr && scope instanceof DrlNameExpr &&
    ((DrlNameExpr) rootExpr).getName().equals(((DrlNameExpr) scope).getName());
```

**Review by mariofusco — APPROVED** [review]
Approved, just 2 very minor changes requests.

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java (hunk 2 above)
Comment (tkobayas) [reply drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java]: Thanks! Fixed.

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java (hunk 3 above)
Comment (tkobayas) [reply drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/PreprocessPhase.java]: Thanks! Fixed.

**Pull request merged**